File: solution/79. Word Search/79.cpp
class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        h = board.size();
        w = board[0].size();
        
        for (int y = 0; y < h; y++)
            for (int x = 0; x < w; x++)
                if (walk(board, word, 0, y, x)) return true;
        
        return false;
    }
    
private:
    int h, w;
    
    bool walk(vector<vector<char>>& board, const string& word, int cur, int y, int x) {
        if (y < 0 || y == h || x < 0 || x == w || word[cur] != board[y][x]) return false;
        if (cur == word.length()-1) return true;
        
        char c = board[y][x];
        board[y][x] = '#';
        bool isFound = walk(board, word, cur+1, y+1, x)
                    || walk(board, word, cur+1, y-1, x)
                    || walk(board, word, cur+1, y, x+1)
                    || walk(board, word, cur+1, y, x-1);
        board[y][x] = c;
    
        return isFound;
    }
};

File: solution/79. Word Search/79.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        m, n = len(board), len(board[0])
        dirs = [0, 1, 0, -1, 0]
        
        def dfs(i, j, k):
            if i < 0 or i >= m or j < 0 or j >= n or board[i][j] != word[k]:
                return False
            if k == len(word) - 1:
                return True
            
            old = board[i][j]
            board[i][j] = '#'
            for d in range(4):
                y = i + dirs[d]
                x = j + dirs[d + 1]
                if dfs(y, x, k + 1):
                    return True
            board[i][j] = old
            return False
        
        for i in range(m):
            for j in range(n):
                if dfs(i, j, 0):
                    return True
        return False
    
